### ISSUE TITLE
XaAES: wind_get(..., WF_TOP, ...)

### ISSUE THREAD

**Issue by mfro0** [freemint/freemint#170]
According to documentation, `wind_get(..., WF_TOP, ...)` with an invalid window handle (e.g. -1) should return the topmost window of a (foreign) application (as well as the AES id of the owner).

It turns out it doesn't. At least not always.

The current code in xa_wind.c (https://github.com/freemint/freemint/blob/ddf829da53264cde4ae019c8a8872c6336cf1eef/xaaes/src.km/xa_wind.c#L1537) 

only returns the window details (owner and handle) if the window has focus (whatever that is, I think it is checking if the application waits for keyboard events). If it does not have focus, `wi_gw1` and `wi_gw2` just keep their previous values.

toshyp does not talk about such constraints.

I would be tempted to remove the `wind_has_focus()`constraint, but that seems to affect toswin2 that appears to rely on this behaviour?

Background is that I'm trying to code an XaAES screenshot application for native ColdFire (as no such thing appears to be available and I'm tired to take photos of my FireBee screen). The XaAES code that calls the snapshot application just provides the information that it is supposed to dump the top window (no coordinates). So there needs to be some functionality to obtain these.

If we'd do an explicit test on a negative window handle (and one >= 0 in the toswin2 case), I would assume the two use cases would not conflict?

**Comment by mfro0** [comment]
fixed the issue, I think. Will do some more tests and a commit later, if all goes well.
![scrap](https://user-images.githubusercontent.com/18102882/81003076-9662c180-8e4a-11ea-83fd-d9d997600f36.jpg)

**Comment by vinriviere** [comment]
Hey, that's beautiful! Really on the FireBee? Wow, all those fonts!

BTW, long ago I complained that TeraDesk didn't always get the focus on startup. IIRC the TosWin2 menu bar was always displayed on startup, that was confusing for users (specially newcomers). Helmut changed something, it had some effect but since then the menu bar sometimes is out of sync with the real focused application. I have no idea it this story is related to your focus issue or not.

**Comment by mikrosk** [comment]
> Background is that I'm trying to code an XaAES screenshot application for native ColdFire (as no such thing appears to be available and I'm tired to take photos of my FireBee screen).

Can't you use the built-in screenshot code in XaAES? `CTRL + ALT + :`, IIRC.

**Comment by mfro0** [comment]
`CTRL + ALT + D`

That's only the 'grab* UI part.
The real work must be done by an external application (see `snapshot` in `xaaes.cnf`)

But I have it working, eventually. Heading for `png` support right now.

**Comment by mfro0** [comment]
> Hey, that's beautiful! Really on the FireBee? Wow, all those fonts!
> 

Sure ;). And if you look closely, there's a hint to your 68kemu as well.

It appears 68kemu needs a little dedication as well. Atari Works works fine on it, but fVDI fonts/GDOS are not recognised. Looks as if `vq_gdos()` return value gets lost somewhere.

**Comment by th-otto** [comment]
>Background is that I'm trying to code an XaAES screenshot application for native ColdFire (as no >such thing appears to be available

Did you try my snapit tool ? http://tho-otto.de/sharedlibs.php It is available for cf too, including png support.
I think i haven't uploaded the source yet, but of course i could do that, too. Most of the work is done by the shared libraries, anyway.
I does not respond to alt-help or some other shortcut, though. It must be accessible as accessory.

**Comment by mfro0** [comment]
> Did you try my snapit tool ? http://tho-otto.de/sharedlibs.php It is available for cf too, including png support.

Not until just now (did not know about its existance, frankly).  Nice. Works fine.

Only it apparently doesn't integrate into XaAES (https://github.com/freemint/freemint/wiki/XaAES#SNAPSHOT)? That way you could snap dropped down menus, etc. as well.

**Comment by th-otto** [comment]
Source is now available at https://github.com/th-otto/SnapIt, if you want do have a look.

>Only it apparently doesn't integrate into XaAES 

You just have to define it as snapshot tool in xaaes.cnf

**Comment by mfro0** [comment]
> Source is now available at https://github.com/th-otto/SnapIt, if you want do have a look.
> 
Ah, great, thank you. I see you had fun with `wind_get(..., WF_TOP, ...)`as well ;). Most likely the same issue I struggle with.

> > Only it apparently doesn't integrate into XaAES
> 
> You just have to define it as snapshot tool in xaaes.cnf

Yes, but not really smooth (or did I do something wrong?). I have to click away XaAES' internal snapshot dialog first to get to the accessory?
Anyway, way better than what I have, I think I drop my attempts and use yours.
Thank you.

I would like to propose the following patch to `xa_wind.c`nevertheless:

    @@ -1534,7 +1534,7 @@ XA_wind_get(int lock, struct xa_client *client, AESPB *pb)
     {
        if (w != root_window)
        {
    -        if (wind_has_focus(w))
    +        if (wind < 0 || wind_has_focus(w))
             {
                DIAG((D_w, client, " -- top window=%d, owner='%s' has focus", w->handle, w->owner->name));
                o[1] = w->handle;

This way, you can always get to the top window (regardless of focus) by providing a negative window handle. Much closer to what the documentation says and does not affect toswin2 behaviour.

**Comment by th-otto** [comment]
> have to click away XaAES' internal snapshot dialog first to get to the accessory?

Oh, ok. I've never used the keyboard shortcut. I think, if the snaphot program is not xaesnap, that dialog should not appear at all.

>I see you had fun with wind_get(..., WF_TOP, ...)as well ;). 

That's possible, didn't remember since it's already quite some time that i wrote the tool.

>I would like to propose the following patch 

I'll look into that. But want to check first what wind_has_focus() actully does. Also i think that if wind < 0, w will be null.

**Comment by mfro0** [comment]
> Also i think that if wind < 0, w will be null.

No, it will be set to `TOP_WINDOW`, just what we are interested in.

**Comment by mikrosk** [comment]
Found this today -- @th-otto do you see any reason why not to commit @mfro0's [change](https://github.com/freemint/freemint/issues/170#issuecomment-623885908)?

**Comment by mikrosk** [comment]
> According to documentation, `wind_get(..., WF_TOP, ...)` with an invalid window handle (e.g. -1) should return the topmost window of a (foreign) application (as well as the AES id of the owner).

@mfro0 I know it has been a long time but ... do you have any proof for this claim? Because I have checked tos.hyp, the compendium, TOS 4.04 source code, AES 4.1 source code and all ignore the handle. So your patch actually introduced a regression.
